### DISCUSSION TITLE
First working use of rtgym for Deep-RL applied to Gran Turismo 1.

### DISCUSSION THREAD

**Discussion by NDR008** [trackmania-rl/tmrl#48] [Show and tell]
First working use of rtgym for Deep-RL via the Rlib framework applied to Gran Turismo on PCSX-Redux emu.. Communicating via TCP sockets, with protobuf for serialisation

Sharing my first working pipeline. :) Major mini-party.

https://youtu.be/zVrhbXNOHCc

**Comment by yannbouteiller** [comment]
Nice progress, thanks for the video :)

Just a detail: at the moment there is no option in rtgym to apply the default action in case of timestep timeout. It is probably possible to implement if you need that, but the philosophy of rtgym is more that timeouts should not happen (when they do, rtgym fires a warning and breaks elasticity of the next timestep, so that the upcoming timesteps stick to their nominal duration instead of being overly compressed to compensate for the overflowing delay).

Instead, the role of the default action has to do with the Markov structure of real-time environments around "reset" transitions. In real-time environments, time cannot be paused, so an action needs to be applied at all time (should this be the "no action" action). So when you call reset(), especially the first time, reset() needs a default action to apply in your environment, because you have to account for the fact that the world is not paused while you are computing your first action. This is where the default action intervenes.

**Comment by NDR008** [comment]
Thanks for the feedback and correcting that point. Will address it in the
next issue.

**Comment by NDR008** [comment]
So after all the celebrations... I feel like I am back to square 1.
I could use the rllib PPO to train my agent to do well in a drag race (it learnt to not zig-zag towards the finish line, and to avoid the brake).

But, PPO only works smoothly with torch, in tf/tf2, things behave weirdly.
SAC just goes nuts after 14 epsidoes, (the timesteps go from 0.05s to .9s).
Not sure what is happening, but my GPU/CPU and main RAM are all OK.
(I got a 3090 on the way, but it should not be the issue really).

I revisited sb3, because their beta version supports gymnasium, but they do not support observation spaces that are tuples, only dicts.
So that felt like another wall.

My research has suggested SAC ought to be the best way... maybe I have no choice but to make a full implementation of tmrl?

**Reply by yannbouteiller** [reply]
When you say the timestep goes to 0.9, you mean rtgym times out repeatedly spamming you with warnings about the timestep being 0.9?

If so, it probably means that rllib/tensorflow is consuming the CPU core used by your python process hosting the rtgym environment.

rllib is not meant for robotics and tmrl with remote training is the optimal solution here, but maybe there is a way of keeping rllib/tensorflow from eating in rtgym's plate.

**Reply by NDR008** [reply]
Yesterday I did some more investigations, but, not really sure if this is a real resource issue (it somehow is but isn't).

First off, PPO on Torch vs TF2 in PPO.
Torch-PPO: Runs well, without CNN, I could train an agent to perform well in a drag-race (0-400m). It eventually learnt to "not brake" and "not touch the steering" (after around 2000 episodes each with a truncated max of 150 steps of 0.1s)

All without CNN:
TF2-PPO: Starts of well, but after around 14 episodes, it starts spamming that it cannot keep the target time-step.
I cannot figure out what's really going on because CPU usage and GPU usage (both for RAM and processing power) remain low (at around 30%).
However, it does feel like it is system resource related because the PC exhibits a small freeze (like 200ms every 5 seconds).
That sort of freeze where the mouse stops moving / and the game freezes (on the emulator side - and the emulator has no relationship to gym, it cannot be effected by anything from the AI). But right after it speeds up again.
Someone wondered if it was a TF library version that was not playing well with rllib.

SAC... zero hope here, whenever I try running Ray's default SAC, after a few epsidoes, the worker stops interacting. So the virtual gamepad remains active on the last button press. Eventually it recovers, but each time the worker seems to freeze, the time to recover seems to get larger.
I wonder if the freeze is occurring when learning/training is done off the replay buffer OR perhaps during a disk-write of the replay buffer.

My understanding, rllib is designed for distributed learning and splitting learner/worker between GPU/CPU (which is why I am thinking the new GPU I ordered might not help anyway).

Either way, it is properly demotivating me, because I cannot find people who can support finding the root cause.

In any case, rllib's default models are not so useful as is for me, because they only have 1D FC models or CNN models (but no mixed modes).
Also using rllib for evaluation/inference seems to be a nightmare (see: https://discuss.ray.io/t/how-to-load-from-check-point-and-call-the-environment/10638)

So... this got me gravitating back to sb3 (main issue seems to be that they rely on dict observations or boxes, no tuples).
I am still nervous about the wrapper solution from below since, rtgym adds the action history as part of the observation - I am not sure if that will make it complicated.

Re-implementing tmrl (but seems like a lot of work, maybe most rewarding, but maybe would require more time than I have available in conjunction with running all my case studies for my dissertation.
Or going for vanilla gym.

One more question: does rtgym do anything "extra" when the time step is within the elastic constraints?
Is it controlling the time-step? If an observation comes too fast (let's say 0.0001s later) and the timestep is set to 0.1s, would it discard that observation?

**Reply by yannbouteiller** [reply]
rtgym simply clocks observation retrieval and action inference, it does not discard anything, it just calls your get_obs_rew_terminated_info in a timely fashion and blocks on step() such that the actual time step sticks to the nominal frequency target that you set.

Your issue is clearly a resource issue, despite the CPU being seemingly under-used. rllib (and most RL frameworks) are used to deal with the normal RL setting where time is paused between time-steps to free resources for all kinds of computation, not with the Real-Time RL setting where time cannot be paused.

You can of course create a non-rtgym environment and use it in rllib as you are doing now with your rtgym environment, but you will still have the same issue rtgym is spamming you warnings about. It will simply be hidden and your training will fail without you understanding why.

My recommendation is rather to use remote training, by setting your worker on another machine and keeping your 3090 for the trainer. Both rllib and tmrl allow you to do this (but not SB3 as far as I know). Then the resources used for training will not get in the way of the worker anymore.

**Reply by NDR008** [reply]
I'm not feeling well today, so I got too many  questions, but will follow up another day. Main ones below:

1. Does it mean / do you suggest, most of the projects like Sophy or the papers I've seen using WRC, or such things rely on:
-Massive horse power
-Remote training
-Frame skipping
-Pausing the environment
?

Might be with implementing a simple gym env to compare.


2. So rtgym doesn't execute step() till elasticity is satisfied?

Maybe a basic question, is the value of rtgym only in inference or also in training?

I guess I still struggle to understand:
https://raw.githubusercontent.com/yannbouteiller/rtgym/main/figures/rt_gym_env.png

There are 2 a's, the green and red - what is the red one?

Also is wait() something that can help me here? (Example to pause emulation?)

**Reply by yannbouteiller** [reply]
Usually they rely on a combination of horse power, remote training and pausing the environment.

rtgym has a threaded structure that executes step() in the background and returns as soon as the observation finishes being captured. Observations are captured at your target frequency (+- elasticity in terms of timewise variance), so if you execute step() twice in less than a timestep, the second call will block until the timestep ends.

**Reply by yannbouteiller** [reply]
Wait(), you shouldn't use that between timesteps. Only rarely when you don't want to capture new steps for a while, but keeping in mind that it breaks the real-time flow of your environment.

**Reply by yannbouteiller** [reply]
The red "a" represents the time during when a call to step() is blocking in the figure you linked

**Reply by NDR008** [reply]
OK, feeling better (health wise). Trying to catch-up with the many fragments of information that I got.
So:
a1(o0,a0) <--- I guess that is action1 based on observation0 and I am unsure what is the a0. I am guessing it is referring to the green a0?

Regarding the trainer issue with rllib, I looked into it, and looks like besides procuring a 2nd PC, I need a linux setup, because clusters are not officially supported on Windows.

Gradually more and more I am thinking that implementing tmrl might be the only way left for me to go...
The ONLY other thing I keep thinking of is... if I re-implement the gamepad method. Instead of using an emulated windows controller, I implement a TCP method, then I can have the emulator run completely on another machine, but I am not sure this would win me much, because I think the issues here are that a worker and trainer run on the same machine, even though I think rllib let's me allocate them to different CPUs/GPUs.

**Reply by yannbouteiller** [reply]
Yeah it is probably tensorflow eating up all your CPU cores including the one running the python process of your Gym environment, or something like that. I am not sure how rllib allocates CPUs.

Implementing a tmrl pipeline is a bit more involved because you will have to implement the memory and the training algorithm, but it will work on windows, it will give you more control than rllib, and following the tutorial should be rather easy.

a0 is the default action, it is applied in the environment while you are computing your first action a1. From a Markov perspective, a1 will thus be applied in a state that is a sum of observation o0 and action a0, which is why both o0 and a0 need to be taken into account when computing a1. The same reasoning applies for all  subsequent actions in the real-time setting.

**Reply by NDR008** [reply]
Actually,I finally spotted the issue, but not sure how to fix it, the issue is that it is maxing out 1 thread... (I have 16, the other 15 are just chilling).
Not sure how to deal with it yet.

**Reply by NDR008** [reply]
Another update... after messing around with ray settings... so far... no spamming of timeouts.
I think the 3090 was... unnecessary for this job other than I now have a gaming PC (even though I am not a PC gamer).

Let's see how far I can get.

**Reply by yannbouteiller** [reply]
The 3090 will come in handy when you try off-policy algorithms like SAC though, then you can try huge batches

**Reply by NDR008** [reply]
Nevermind... I thought I got SAC working, but it was just A2C...

**Reply by NDR008** [reply]
Is there a way for me to remove the action history from rtgym?

**Reply by NDR008** [reply]
"My recommendation is rather to use remote training, by setting your worker on another machine and keeping your 3090 for the trainer. Both rllib and tmrl allow you to do this"

Just in case - do you know how to set the worker on a remote machine?

**Reply by NDR008** [reply]
I think this is what I need to look into:
https://docs.ray.io/en/latest/rllib/rllib-env.html#Hierarchical:~:text=hierarchical_training.py.-,External,-Agents%20and%20Applications

I think the issue is not, CPU/GPU power, but that rllib is design to roll-up the environment into its workers, and it assume that the worker controls step(), but when there is a long delay between step() calls, I think rtgym still complains.
rllib is not aware that the environment is still stepping.
(And since the virtual gamepad remains in the last action-state, it looks like hanging, but isn't).

But it seems by using the client/server examples mentioned here might be my way: https://docs.ray.io/en/latest/rllib/rllib-env.html#external-agents-and-applications:~:text=hierarchical_training.py.-,External,-Agents%20and%20Applications

**Reply by NDR008** [reply]
> Is there a way for me to remove the action history from rtgym?

Sorry does rtgym work if I set the action buffer length to 0?

**Reply by yannbouteiller** [reply]
You can remove the action buffer from rtgym by setting `"act_in_obs"` to `False` in the configuration dictionary.

Are you sure you want to do that though? It only makes sense if you are using an RNN or another way of handling real-time non-Markovness.

**Comment by NDR008** [comment]
Do you think it is possible to make a simple wrapper to rtgym to convert the observation tuple to a dict?

**Reply by yannbouteiller** [reply]
Sure, gymnasium has a [Wrapper](https://gymnasium.farama.org/api/wrappers/#gymnasium-wrapper) class that you can use to transform the observation space of any gymnasium environment, including rtgym.

**Comment by NDR008** [comment]
I just want to debug a few issues and tes.t some others. My plan was to
fully stack the observation. I'm not sure why actions ought to be part of
the observation to make it more Markovian...

**Reply by yannbouteiller** [reply]
Because `rtgym` environments are real-time: you don't pause the simulator (unless you are performing computations between atomic game frames), thus stuff happens in the simulator while you are computing a new action, and thus you need the applied action to be part of the observation used to compute this new action, otherwise this observation does not satisfy the Markov property.

**Reply by NDR008** [reply]
I understand what you are saying in "gut-feeling" sense, but not in a mathematical or theoretical sense.
Similarly, I guess the size of action history, is probably an empirically set parameter?

**Reply by yannbouteiller** [reply]
No it is defined by the duration of your maximum delay (probably 2 time steps unless there is additional delay happening in the game after you sent the gamepad command). You can google "reinforcement learning with random delays" if you want to understand this in mathematical terms, but put simply, because time is not paused while you compute an action or even capture an observation, the action that your NN computes will be applied in a future state that is not the state it actually sees. To predict this future state, the NN needs to observe the actions that will lead to it.
